DDP-7039 - implement user full delete service

evgeniipr opened the pull request:
## Context

Implement a **user full delete** service; implement a new PubSub task of type USER_DELETE  (in order to be possible to delete user through PubSub).

**Basic changes:**
-  implemented service `UserFullDeleteService` providing user's all data deletion (it utilizes a service `UserDeleteService` which used to delete a user through a REST service;
- added a new PubSubTask type: `USER_DELETE` and implemented `UserDeleteProcessor` for an incoming PubSub task processing (user deletion).

## Checklist

- [x] I have labeled the type of changes involved using the `C-*` labels.
- [x] I have assessed potential risks and labeled using the `R-*` labels.
- [x] I have considered error handling and alerts, and added `L-*` labels as needed.
- [x] I have considered security and privacy, and added `I-*` labels as needed
- [x] I have analyzed my changes for stability, fault tolerance, graceful degradation, performance bottlenecks and written a brief summary in this PR.
- [x] If applicable, I have discussed the analytics needs at both a platform and study level with Product and instrumented code accordingly.
- [x] If applicable, my UI/UX changes have passed muster with Product/Design via an over-the-shoulder review, screenshots, etc.

_If unsure or need help with any of the above items, add the `help wanted` label. For items that starts with `If applicable`, if it is not applicable, check it off and add `n/a` in front._

## FUD Score

_Overall, how are you feeling about these changes?_

- [x] :relaxed: All good, business as usual!
- [ ] :sweat_smile: There might be some issues here
- [ ] :scream: I'm sweaty and nervous

## How do we demo these changes?

_How does one observe these changes in a deployed system? Note that **user visible** encompasses many personas--not just patients and study staff, but also ops duty, your fellow devs, compliance, etc._

- [x] They are user-visible in dev as a regular user journey and require no additional instructions.
- [ ] Getting dev into a state where this is user-visible requires some tech fiddling. I have documented these steps in the related ticket.
- [ ] Requires other features before it's human visible. I have documented the blocking issues in jira.
- [ ] I have no idea how to demo this. Please help me!

## Testing

- [x] I have written automated positive tests
- [x] I have written automated negative tests
- [ ] I have written zero automated tests but have poked around locally to verify proper functionality
- [ ] The jira ticket has acceptance criteria and QA has the needed information to test changes

## Release

- [x] These changes require no special release procedures--just code!
- [ ] Releasing these changes requires special handling and I have documented the special procedures in the release plan document

Commit:
DDP-7039 - implement user full delete service; implement a new PubSub task type USER_DELETE - in order to be possible to delete user through PubSub

evgeniipr commented on pepper-apis/src/main/java/org/broadinstitute/ddp/db/dao/KitTypeDao.java:
```diff
@@ -27,7 +27,7 @@
     @SqlQuery("SELECT kit_type_id, name FROM kit_type WHERE name = :kitTypeName")
     @RegisterRowMapper(KitTypeRowMapper.class)
-    Optional<KitType> getKitTypeByName(@Bind String kitTypeName);
+    Optional<KitType> getKitTypeByName(@Bind("kitTypeName") String kitTypeName);
```
Fixed text

evgeniipr commented on pepper-apis/src/main/java/org/broadinstitute/ddp/db/dto/EnrollmentStatusDto.java:
```diff
@@ -12,8 +13,15 @@
+            @ColumnName("study_guid") String studyGuid,
+            @ColumnName("enrollment_status") EnrollmentStatusType enrollmentStatus,
+            @ColumnName("valid_from_millis") long validFromMillis,
+            @ColumnName("valid_to_millis") Long validToMillis) {
```
Fixed test

evgeniipr commented on pepper-apis/src/main/java/org/broadinstitute/ddp/event/publish/pubsub/TaskPubSubPublisher.java:
```diff
@@ -45,9 +45,9 @@ private PubsubMessage createPubSubMessage(String taskType, String eventPayload,
-                ATTR_PARTICIPANT_GUID, participantGuid
+                ATTR_NAME__TASK_TYPE, taskType,
+                ATTR_NAME__STUDY_GUID, studyGuid,
+                ATTR_NAME__PARTICIPANT_GUID, participantGuid
```
Constants' names refactoring.

evgeniipr commented on pepper-apis/src/main/java/org/broadinstitute/ddp/event/pubsubtask/api/PubSubTask.java:
```diff
@@ -11,7 +11,13 @@
+    /** PubSubTask standard attribute: holds a study GUID */
+    public static final String ATTR_NAME__STUDY_GUID = "studyGuid";
+    /** PubSubTask standard attribute: holds a participant (user) GUID */
+    public static final String ATTR_NAME__PARTICIPANT_GUID = "participantGuid";
```
Put all common (used by all tasks) attributes to `PubSubTask`.

Commit:
DDP-7039 - remove unused fields; refactor classes structure

evgeniipr commented on pepper-apis/src/main/java/org/broadinstitute/ddp/event/pubsubtask/api/PubSubTaskResult.java:
```diff
@@ -15,10 +15,10 @@
-    public static final String FIELD_RESULT_TYPE = "resultType";
-    public static final String FIELD_ERROR_MESSAGE = "errorMessage";
+    public static final String PAYLOAD_FIELD__RESULT_TYPE = "resultType";
+    public static final String PAYLOAD_FIELD__ERROR_MESSAGE = "errorMessage";
```
Renamed constants: make it more readable.

evgeniipr commented on pepper-apis/src/main/java/org/broadinstitute/ddp/event/pubsubtask/impl/PubSubTaskProcessorFactoryImpl.java:
```diff
@@ -18,5 +19,9 @@ protected void registerPubSubTaskProcessors() {
+        registerPubSubTaskProcessor(
+                TASK_TYPE__USER_DELETE,
+                new UserDeleteProcessor()
+        );
```
Register new PubSubTask type

Commit:
DDP-7039 - prevent deletion of users having references to revisoon history; improve log and error messages; remove unused parameters

MocanaAtBroad commented on pepper-apis/src/main/java/org/broadinstitute/ddp/service/userdelete/UserService.java:
```diff
@@ -1,4 +1,4 @@
-package org.broadinstitute.ddp.service;
+package org.broadinstitute.ddp.service.userdelete;
```
Don't think this is a good name for a package. I suggest we stick with original location.

evgeniipr replied:
Fixed

MocanaAtBroad commented on pepper-apis/src/main/java/org/broadinstitute/ddp/service/userdelete/UserFullDeleteService.java:
```diff
@@ -0,0 +1,147 @@
+    }
+
+    private void deleteUserSteps(Handle handle, User user) throws IOException {
+        deleteAuth0User(handle, user);
```
Since this in Auth0, I would prefer that this be the last thing we remove: if the full database transaction removing the user succeeds then we can proceed to remove from Auth0

evgeniipr replied:
Put the auth0 deletion at the end.

MocanaAtBroad commented:
I am confused about having a UserFullDeleteService and a UserService.deleteUser method. Why can't we just have one method?

MocanaAtBroad commented on pepper-apis/src/main/java/org/broadinstitute/ddp/service/userdelete/UserFullDeleteService.java:
```diff
@@ -0,0 +1,147 @@
+        deleteUserStudyLegacyData(handle, user);
+        deleteUserAnnouncement(handle, user);
+        deleteActivityInstanceStatus(handle, user);
+        deleteAnswersByOperator(handle, user);
```
Are we sure that we want to delete answers only by operator id and not  participant id?
I am thinking of case whether a parent signs up child and we need to delete the childs data.
Then earlier check of not having any governed users will pass (the child does not govern any users).
But the deltreAnswersByOperator is not going to find the child's data because child is not operator.
Am I missing something? does userService.deleteUser delete the child's data?

ssettipalli replied:
Data is associated with child. 
I don't think we can delete a parent/proxy user without deleting ALL dependents/governed-users.
We should check if current user has any governed users first and exit with response that user has governed users and cannot be deleted until ALL governed users are deleted first which is what I believe we are doing here.

May be we should go with the assumption that ALL governed users will be deleted one at a time before attempting to delete the proxy/parent/guardian.

evgeniipr replied:
> Are we sure that we want to delete answers only by operator id and not participant id?

@MocanaAtBroad this is according to one of deletion statements from a confluence page you referred in the ticket:
`delete from answer where operator_user_id in (select user_id from delete_users);`

MocanaAtBroad replied:
OK. Please verify this, but what I am seeing is that this is not necessary. We block deleting users that have governed users already. And, I believe, the only case where the participant id is going to be different from the operator id is case when THERE ARE governed users. You delete the participants data a little later.
I don't think this line does any harm, but makes me nervous having it there. If my analysis is correct (please check my logic) then let's drop the deletion by operator id.

ssettipalli commented on pepper-apis/src/main/java/org/broadinstitute/ddp/service/userdelete/UserFullDeleteService.java:
```diff
@@ -0,0 +1,147 @@
+        deleteActivityInstanceStatus(handle, user);
+        deleteAnswersByOperator(handle, user);
+
+        userService.deleteUser(handle, user);
```
We need a step to delete the user from elastic indices. user, participant and participant_structured indices.

evgeniipr replied:
These ES deletion steps are implemented in UserDelete service (was implemented earlier by Yury Pavlov) and which UserFullDeleteService uses.. But I agree with @MocanaAtBroad that better to define only one service for both kind of deletions.. I need to reimplement this service.

evgeniipr replied:
Re-implemented services: created only one service - UserDeleteService.

evgeniipr commented:
> I am confused about having a UserFullDeleteService and a UserService.deleteUser method. Why can't we just have one method?

Initially I didn't wanted to touch the existing deletion method in UserService (which used by Rest endpoint "delete user"). 
Or you suggest to use only one service and add some parameter (in constructor) which defines a level of deletion (for example delete/not delete auth0)...? Seems this is good approach. I like it.

Commit:
DDP-7039 - refactor code: create single class UserDeleteService which used for both kinds of user deletion

Commit:
DDP-7039 - improve logging

evgeniipr commented on pepper-apis/src/main/java/org/broadinstitute/ddp/service/UserDeleteService.java:
```diff
@@ -0,0 +1,367 @@
+ *  <li>send data-sync request.</li>
+ * </ul>
+ */
+public class UserDeleteService {
```
Instead of 2 services (UserFullDeleteService and UserService) only one service for user deletion is created: UserDeleteService.

evgeniipr commented:
@MocanaAtBroad I re-implemented services: created only one service - UserDeleteService.

evgeniipr committed:
Merge branch 'develop' into DDP-7039-user-full-delete

ssettipalli commented on pepper-apis/src/main/java/org/broadinstitute/ddp/service/UserDeleteService.java:
```diff
@@ -0,0 +1,367 @@
+            deleteKitRequests(handle, user, logInfo);
+            deleteUserStudyLegacyData(handle, user, logInfo);
+            deleteUserAnnouncement(handle, user, logInfo);
+            deleteActivityInstanceStatus(handle, user, logInfo);
```
SimpleDelete doesn't have activityInstanceStatus ? 
User with Activity Instance will have activity Instance status, right !!

evgeniipr replied:
activity_instance_status is deleted during simpleDelete too. 
This ` deleteActivityInstanceStatus(handle, user, logInfo);` is additional deletion by operator_id.
I renamed it to `deleteActivityInstanceStatusByOperator`.
In the manual delete instruction (https://broadinstitute.atlassian.net/wiki/spaces/DDP/pages/856260664/User+Delete) there are 2 rows of status deletion:
```
delete from activity_instance_status where activity_instance_id in (select activity_instance_id from activity_instance where participant_id in (select user_id from delete_users));delete from activity_instance_substitution where activity_instance_id in (select activity_instance_id from activity_instance where participant_id in (select user_id from delete_users));

delete from activity_instance_status where operator_id in (select user_id from delete_users);
```
The 1st one called during both simpleDelete and fullDelete.
The 2nd one - only during fullDelete (at least in the initial logic of simple delete this statement was not executed..)

Btw, seems I need to re-create the PR: I created this new class UserDeleteService not quite correctly: added it as a new class and deleted old class `UserService`. Better to rename `UserService` to `UserDeleteService` in order to preserve the history of changes...

Commit:
DDP-7039 - improve logging: add user info to log message

Commit:
Merge branch 'DDP-7039-user-full-delete' of https://github.com/broadinstitute/ddp-study-server into DDP-7039-user-full-delete

Commit:
DDP-7039 - fix test

Commit:
DDP-7039 - added additional fields to the user delete task: whoDeleted, comment; fixed test

MocanaAtBroad commented on pepper-apis/src/main/java/org/broadinstitute/ddp/service/UserDeleteService.java:
```diff
@@ -0,0 +1,381 @@
+ *  <li>delete ES data;</li>
+ *  <li>delete auth0;</li>
+ *  <li>send data-sync request.</li>
+ * </ul>
```
Let's drop the steps from comment. Hopefully the code is clear enough to illustrate what the steps are. And this method will probably have to be updated as schema evolves.

evgeniipr replied:
Removed comments.

MocanaAtBroad commented on pepper-apis/src/main/java/org/broadinstitute/ddp/service/UserDeleteService.java:
```diff
@@ -0,0 +1,381 @@
+     * Used to delete a user which still not registered in auth0-system, not governed by others than a user
+     * who deletes this user, not having status ENROLLED.
+     */
+    public void simpleDelete(Handle handle, User user, String whoDeleted, String comment) throws IOException {
```
A full javadoc comment would be useful here, particularly explaining what this arguments mean and do. e.g. whoDeleted: is this just a comment string or an identifier like a guid or an email? I can't tell.

evgeniipr replied:
Added comments for parameters.

MocanaAtBroad commented on pepper-apis/src/main/java/org/broadinstitute/ddp/service/UserDeleteService.java:
```diff
@@ -0,0 +1,381 @@
+        }
+    }
+
+    private LogInfo deleteUserSteps(Handle handle, User user, boolean fullDelete) throws IOException {
```
Would rename to simply "deleteUser". The word "Steps" is misleading: I originally thought we were deletinguser  steps from the database.

evgeniipr replied:
Renamed

MocanaAtBroad commented on pepper-apis/src/main/java/org/broadinstitute/ddp/service/UserDeleteService.java:
```diff
@@ -0,0 +1,381 @@
+            deleteUserStudyLegacyData(handle, user, logInfo);
+            deleteUserAnnouncement(handle, user, logInfo);
+            deleteActivityInstanceStatusByOperator(handle, user, logInfo);
+            deleteAnswersByOperator(handle, user, logInfo);
```
Hmm. Made comment about not needing this deleteAnswersByOperator elsewhere. github can be confusing....

MocanaAtBroad commented on pepper-apis/src/main/java/org/broadinstitute/ddp/service/UserDeleteService.java:
```diff
@@ -0,0 +1,381 @@
+        deleteMedicalProvider(handle, user, logInfo);
+        deleteEnrollmentStatuses(handle, user, logInfo);
+        deleteUserAddresses(handle, user, logInfo);
+        deleteAnswersAndActivityInstances(handle, user, logInfo);
```
Might as well rename to deletePartcipantAnswersAndActivityInstances

evgeniipr replied:
Renamed

MocanaAtBroad commented on pepper-apis/src/main/java/org/broadinstitute/ddp/service/UserDeleteService.java:
```diff
@@ -0,0 +1,381 @@
+    }
+
+    private void deleteKitRequests(Handle handle, User user, LogInfo logInfo) {
+        logInfo.add("kit_request");
```
Collecting the tables touched by deletion into a LogInfo object is an interesting idea, but I don't think it is particularly useful to the caller of the method. There is nothing that the caller can do with the information. I would rather write to the real log the operations as we do them and see how many rows of each kind have been removed. And if something goes wrong with a step of the deletion, then a developer with access to logs can see how far it got from the log. The LogInfo is not even going to be returned in case of failure.

evgeniipr replied:
Replaced this logging approach to separate log calls before each deletion step.

MocanaAtBroad commented:
Could you double check that we are indeed getting everything? This is the list of columns that have a foreign key reference to the `user` table. At bottom of this comment is the query I used to generate this.
I am not so sure about table `file_upload` being properly deleted.


|TABLE_NAME                            |COLUMN_NAME          |
|--------------------------------------|---------------------|
|activity_instance                     |participant_id       |
|activity_instance_creation_mutex      |participant_id       |
|activity_instance_status              |operator_id          |
|age_up_candidate                      |participant_user_id  |
|answer                                |operator_user_id     |
|data_sync_request                     |user_id              |
|default_mailing_address               |participant_user_id  |
|event_configuration_occurrence_counter|participant_user_id  |
|file_upload                           |operator_user_id     |
|file_upload                           |participant_user_id  |
|invitation                            |user_id              |
|kit_request                           |participant_user_id  |
|kit_schedule_record                   |participant_user_id  |
|mailing_address                       |operator_user_id     |
|mailing_address                       |participant_user_id  |
|queued_event                          |operator_user_id     |
|queued_event                          |participant_user_id  |
|revision                              |changed_by_user_id   |
|revision                              |terminated_by_user_id|
|study_admin                           |user_id              |
|study_exit_request                    |user_id              |
|temp_mailing_address                  |operator_user_id     |
|temp_mailing_address                  |participant_user_id  |
|user_announcement                     |user_id              |
|user_governance                       |operator_user_id     |
|user_governance                       |participant_user_id  |
|user_legacy_info                      |participant_user_id  |
|user_medical_provider                 |user_id              |
|user_profile                          |user_id              |
|user_study_enrollment                 |user_id              |
|user_study_legacy_data                |user_id              |

```sql
SELECT DISTINCT
    TABLE_NAME, COLUMN_NAME
FROM
    INFORMATION_SCHEMA.KEY_COLUMN_USAGE
WHERE
    REFERENCED_TABLE_NAME = 'user'
ORDER BY TABLE_NAME , COLUMN_NAME;
```

evgeniipr commented:
> Could you double check that we are indeed getting everything?

Actually not everything: I checked that still needs to delete from: activity_instance_creation_mutex, file_upload, invitation, kit_schedule_record, study_admin, study_exit_request, user_legacy_info. 
Added deletion.
I didn't added deletion of `revision` because seems we decided that if user has reference to revisions then we cannot delete such user.

evgeniipr commented:
> OK. Please verify this, but what I am seeing is that this is not necessary. We block deleting users that have governed users already. And, I believe, the only case where the participant id is going to be different from the operator id is case when THERE ARE governed users. You delete the participants data a little later.
I don't think this line does any harm, but makes me nervous having it there. If my analysis is correct (please check my logic) then let's drop the deletion by operator id.

I agree that this delete by operator is not needed: anyway all answers are deleted by activity_instance_id's. 
So, I removed this delete.

Commit:
DDP-7095 - add deletion of user data from all possible tables; change logging approach

MocanaAtBroad commented on pepper-apis/src/main/java/org/broadinstitute/ddp/service/UserDeleteService.java:
```diff
@@ -0,0 +1,416 @@
+            deleteUserStudyLegacyData(handle, user);
+            deleteUserAnnouncement(handle, user);
+            deleteActivityInstanceStatusByOperator(handle, user);
+            deleteFileUpload(handle, user);
```
I would classify fileupload deletion as part of removing activity data and therefore make it part of "normal" deletion.

evgeniipr replied:
Moved to "normal" deletion.

Commit:
DDP-7039 - fix ordering of user data deletion; throw an error in case if ES indices deletion failed (during user full deletion)

Commit:
DDP-7039 - remove 'whoDeleted' from PubSub message payload and added an attribute 'operatorGuid' (optional)

evgeniipr commented:
@MocanaAtBroad , @ssettipalli I updated PubSub task "USER_DELETE" format: removed from payload a property "whoDeleted" and added to attributes an attribute "operatorGuid" (optional).

Commit:
Merge branch 'develop' into DDP-7039-user-full-delete

Commit:
DDP-7039 - fix comments
